### DISCUSSION TITLE
One ping/pong failure will make the websocket disconnect, is that too sensitive ?

### DISCUSSION THREAD

**Discussion by qiulang** [socketio/socket.io#4161] [Q&A] [answered]
Correct me if I am wrong I think that one ping/pong failure will make the websocket disconnect. Is this too sensitive, e.g. why doesn't the heartbeat mechanism add some re-try before it marks the websocket disconnect ?

I know engine.io has reconnect option like reconnectionDelay & randomizationFactor (their meanings are obvious). I assume after a ping/pong failure socket.io will use a new websocket to reconnect. But I can't find a clear document to explain this reconnect process. Can someone shed some light on this ?

**Comment by darrachequesne** [comment, marked as answer]
> why doesn't the heartbeat mechanism add some re-try before it marks the websocket disconnect ?

That's a good question. If a ping/pong fails, then the WebSocket connection is surely in an invalid state, and thus retrying will only delay the reconnection.

> I assume after a ping/pong failure socket.io will use a new websocket to reconnect

Yes, that's right. The codebase is split in two projects:

- Engine.IO: the `engine.io-client` package provides a WebSocket-like object, which falls back to HTTP long-polling (but no reconnection, similarly to the WebSocket API). The heartbeat mechanism happens here, and the connection gets closed if the heartbeat fails.

Source: https://github.com/socketio/engine.io-client/

- Socket.IO: the `socket.io-client` package provides some syntactic sugar over Engine.IO, with an API inspired from the Node.js EventEmitter (`socket.on() / socket.emit()`). The reconnection is handled here: if the underlying Engine.IO client gets closed, then it creates another one.

Source: https://github.com/socketio/socket.io-client/

**Reply by qiulang** [reply]
Thanks for answering my question! 

One of reasons I asked my question is that through our socket.io server logs I found some client-side disconnect events were missing. I searched socket.io issues found this one https://github.com/socketio/socket.io/issues/635. It was closed in 2017 but others commented they still hit the problem. I will open [another discussion](https://github.com/socketio/socket.io/discussions/4169) for that.
